Tutorial for shape optimization of a multilayer stack (#2862)
NanoComp/meep: doc/docs/Python_Tutorials/Adjoint_Solver.md

The stack consists of two materials of alternating refractive index $n_A$ = 1.3 and $n_B$ = 1.0. The layers are arranged as $n_A$, $n_B$, $n_A$, $n_B$, ..., $n_A$. The semi-infinite regions to the left and right of the stack are vacuum.

![](../images/multilayer_stack_shape_opt.png#center)

A reference design to compare our results against is a [quarter-wavelength stack](https://en.wikipedia.org/wiki/Distributed_Bragg_reflector). The mean wavelength of $\lambda_1$ and $\lambda_2$ is $\lambda = 1.0 μm for which the quarter-wavelength layer thicknesses are $\lambda / (4 n_A)$ = 0.19 μm and $\lambda / (4 n_B)$ = 0.25 μm. These values are used to specify upper and lower bounds for the layer thicknesses. This is important given the non-convex nature of this particular design problem.

> The worst-case optimization is implemented using the [epigraph formulation](https://nlopt.readthedocs.io/en/latest/NLopt_Introduction/#equivalent-formulations-of-optimization-problems) and the Conservative Convex Separable Approximation (CCSA) algorithm. The optimization is run ten times with random initial designs from which the local optima with the smallest objective value is chosen. For the run involving the best design, a plot of the objective function vs. iteration number for the two wavelengths is shown below. Also included in this plot is the epigraph variable. These results demonstrate that the stack is being optimized for $\lambda$ = 0.95 μm and *not* $\lambda$ = 1.05 μm. We observed this trend for various local optima. The optimizer converges to the local optima using 13 iterations. The nine layer thicknesses (μm) of this design are: 0.1822, 0.2369, 0.1822, 0.2369, 0.1822, 0.2369, 0.1910, 0.2504, 0.1931. Since these are 1D simulations, the runtime for each design iteration is generally fast (about ten seconds using one core of an Intel Xeon i7-7700K CPU @ 4.20GHz).

Reviewer: ```suggestion
The worst-case optimization is implemented using the [epigraph formulation](https://nlopt.readthedocs.io/en/latest/NLopt_Introduction/#equivalent-formulations-of-optimization-problems) and the Conservative Convex Separable Approximation (CCSA) algorithm. The optimization is run ten times with random initial designs from which the local optimum with the smallest objective value is chosen. For the run involving the best design, a plot of the objective function vs. iteration number for the two wavelengths is shown below. Also included in this plot is the epigraph variable. These results demonstrate that the performance of the stack is limited by that of the shorter wavelength $\lambda$ = 0.95 μm, so the $\lambda$ = 1.05 μm performance does not directly affect the result. We observed this trend for various local optima. The optimizer converges to the local optimum after 13 iterations. The nine layer thicknesses of this design are (in μm): 0.1822, 0.2369, 0.1822, 0.2369, 0.1822, 0.2369, 0.1910, 0.2504, 0.1931. Since these are 1D simulations, the runtime for each design iteration is generally fast (about ten seconds using one core of an Intel Xeon i7-7700K CPU @ 4.20GHz).
```

The stack consists of two materials of alternating refractive index $n_A$ = 1.3 and $n_B$ = 1.0. The layers are arranged as $n_A$, $n_B$, $n_A$, $n_B$, ..., $n_A$. The semi-infinite regions to the left and right of the stack are vacuum.

![](../images/multilayer_stack_shape_opt.png#center)

A reference design to compare our results against is a [quarter-wavelength stack](https://en.wikipedia.org/wiki/Distributed_Bragg_reflector). The mean wavelength of $\lambda_1$ and $\lambda_2$ is $\lambda = 1.0 μm for which the quarter-wavelength layer thicknesses are $\lambda / (4 n_A)$ = 0.19 μm and $\lambda / (4 n_B)$ = 0.25 μm. These values are used to specify upper and lower bounds for the layer thicknesses. This is important given the non-convex nature of this particular design problem.

> 

Reviewer: ```suggestion
```

    )
```

Shape Optimization of a Multilayer Stack
----------------------------------------

> We extend the demonstration of the shape derivative from the previous tutorial to perform shape optimization of a multilayer stack over a broad bandwidth. The 1D design problem is shown in the schematic below and involves finding the layer thicknesses for a fixed number of layers (9) which minimize the largest transmittance (or alternatively, maximize the smallest reflectance) at two wavelengths: $\lambda_1$ = 0.95 μm and $\lambda_2$ = 1.05 μm. However, rather than use the transmittance as the objective function, we use the *field decay* via (the logarithm of) the integral of the fields in the stack ($\int |E_x|^2$). This provides more information to the optimizer which should produce better results.

Reviewer: Maybe we should be minimizing transmittance now that it's not too small.

Reviewer: I'm worried that we are saying something that is false here.

> We extend the demonstration of the shape derivative from the previous tutorial to perform shape optimization of a multilayer stack over a broad bandwidth. The 1D design problem is shown in the schematic below and involves finding the layer thicknesses for a fixed number of layers (9) which minimize the largest transmittance (or alternatively, maximize the smallest reflectance) at two wavelengths: $\lambda_1$ = 0.95 μm and $\lambda_2$ = 1.05 μm. However, rather than use the transmittance as the objective function, we use the *field decay* via (the logarithm of) the integral of the fields in the stack ($\int |E_x|^2$). This provides more information to the optimizer which should produce better results.

Reviewer: ```suggestion
We extend the demonstration of the shape derivative from the previous tutorial to perform shape optimization of a multilayer stack over a broad bandwidth. The 1D design problem is shown in the schematic below and involves finding the layer thicknesses for a fixed number of layers (9) which minimize the integrated field intensity $\int |E_x|^2$, which demonstrates a `FourierFields` objective function.  (This is equivalent to minimizing absorption if $\varepsilon$ had a small imaginary part, and is related but not precisely equivalent to minimizing transmission or maximizing reflection.)  In particular, we minimize the worst case (largest) of the intensities at two wavelengths: $\lambda_1$ = 0.95 μm and $\lambda_2$ = 1.05 μm, to roughly emulate a broadband problem.
```

The stack consists of two materials of alternating refractive index $n_A$ = 1.3 and $n_B$ = 1.0. The layers are arranged as $n_A$, $n_B$, $n_A$, $n_B$, ..., $n_A$. The semi-infinite regions to the left and right of the stack are vacuum.

![](../images/multilayer_stack_shape_opt.png#center)

> A reference design to compare our results against is a [quarter-wavelength stack](https://en.wikipedia.org/wiki/Distributed_Bragg_reflector). The mean frequency of $\lambda_1$ and $\lambda_2$ is $\lambda \approx$ 1.0 μm for which the quarter-wavelength layer thicknesses are approximately $\lambda / (4 n_A)$ = 0.19 μm and $\lambda / (4 n_B)$ = 0.25 μm. These values are used to specify upper and lower bounds for the layer thicknesses. This is important given the non-convex nature of this particular design problem.

Reviewer: ```suggestion
A reference design to compare our results against is a [quarter-wavelength stack](https://en.wikipedia.org/wiki/Distributed_Bragg_reflector). The mean wavelength of $\lambda_1$ and $\lambda_2$ is $\lambda = 1.0 μm for which the quarter-wavelength layer thicknesses are $\lambda / (4 n_A)$ = 0.19 μm and $\lambda / (4 n_B)$ = 0.25 μm. These values are used to specify upper and lower bounds for the layer thicknesses. This is important given the non-convex nature of this particular design problem.
```

> (For several layers and a single wavelength, the optimizer can easily find a non-quarter-wavelength structure with very low transmission, and once it gets to that point the gradients will be almost zero so probably it will converge very slowly towards a quarter-wavelength solution because the optimizer is generally very slow to try to reduce transmittance from 1e-3 to 1e-6, for example.)

Reviewer: ```suggestion
```
